doc/static/downloads/build_acrn_ovmf.sh

```diff
@@ -0,0 +1,220 @@
+#
+# For more information, ./build_acrn_ovmf.sh -h
+#
+
+gop_bin_dir="./gop"
+docker_image_name="ubuntu:ovmf.16.04"
+proxy_conf="proxy.conf"
+acrn_ver="latest"
+
+if [ ! -x "$(command -v docker)" ]; then
+    echo "Install Docker first:"
+    echo "If you are using Ubuntu, you can refer to: https://docs.docker.com/engine/install/ubuntu/"
+    exit
+fi
+
+if [ ! -d "${gop_bin_dir}" ]; then
+    mkdir ${gop_bin_dir}
+    echo "Copy IntelGopDriver.efi and Vbt.bin to ${gop_bin_dir}"
+    exit
+fi
+
+if [ ! -f "${gop_bin_dir}/IntelGopDriver.efi" ]; then
+    echo "Copy IntelGopDriver.efi to ${gop_bin_dir}"
+    exit
+fi
+
+if [ ! -f "${gop_bin_dir}/Vbt.bin" ]; then
+    echo "Copy Vbt.bin to ${gop_bin_dir}"
+    exit
+fi
+
+if [ ! -f "${proxy_conf}" ]; then
+    touch "${proxy_conf}"
+fi
+
+usage()
+{
+    echo "$0 [-v ver] [-i] [-s] [-h]"
+    echo "  -v ver: The release version of ACRN, e.g. 2.3"
+    echo "  -i:     Delete the existing docker image ${docker_image_name} and re-create it"
+    echo "  -s:     Delete the existing acrn-edk2 source code and re-download/re-patch it"
+    echo "  -h:     Show this help"
+    exit
+}
+
+re_download=0
+re_create_image=0
+
+while getopts "hisv:" opt
+do
+    case "${opt}" in
+        h)
+            usage
+            ;;
+        i)
+            re_create_image=1
+            ;;
+        s)
+            re_download=1
+            ;;
+        v)
+            acrn_ver=${OPTARG}
+            ;;
+        ?)
+            echo "${OPTARG}"
+            ;;
+    esac
+done
+shift $((OPTIND-1))
+
+if [[ "${re_create_image}" -eq 1 ]]; then
+    if [[ "$(docker images -q ${docker_image_name} 2> /dev/null)" != "" ]]; then
+        echo "===================================================================="
+        echo "Deleting the old Docker image ${docker_image_name}  ..."
+        echo "===================================================================="
+        docker image rm -f "${docker_image_name}"
+    fi
+fi
+
+if [[ "${re_download}" -eq 1 ]]; then
+    echo "===================================================================="
+    echo "Deleting the old acrn-edk2 source code ..."
+    echo "===================================================================="
+    sudo rm -rf acrn-edk2
+fi
+
+create_acrn_edk2_workspace()
+{
+    echo "===================================================================="
+    echo "Downloading & patching acrn_edk2 source code ..."
+    echo "===================================================================="
+
+    [ -d acrn-edk2 ] && sudo rm -rf acrn-edk2
+
+    git clone https://github.com/projectacrn/acrn-edk2.git
+    if [ $? -ne 0 ]; then
+        echo "git clone acrn-edk2 failed"
+        return 1
+    fi
+
+    cd acrn-edk2
+    git submodule update --init CryptoPkg/Library/OpensslLib/openssl
+    if [ $? -ne 0 ]; then
+        echo "git submodule acrn-edk2 failed"
+        return 1
+    fi
+
+    if [ "${acrn_ver}" != "latest" ]; then
```

Comment:
I think it would be better to put this section before the submodule initialization. It does not matter too much at the moment as we always point at master in the submodule but that may change in the future and so it would be better to make sure the submodule update picks up the version specified in tagged `acrn-edk2` source code.

Reply:
My understanding is that when we change to a tag in the acrn-edk2 repo, the sub-module will be checked out to the according commit id of that tag.

Reply:
As far as I know, this is not the default behaviour. You need to add `--recurse-submodules` to the `checkout` command if you want that. It's actually a very good option, even better than moving this block.

https://git-scm.com/docs/git-checkout#Documentation/git-checkout.txt---recurse-submodules

Reply:
https://git-scm.com/book/en/v2/Git-Tools-Submodules

The DbConnector directory is there, but empty. You must run two commands: git submodule init to initialize your local configuration file, and git submodule update to fetch all the data from that project and check out the appropriate commit listed in your superproject:

There is another way to do this which is a little simpler, however. If you pass --recurse-submodules to the git clone command, it will automatically initialize and update each submodule in the repository, including nested submodules if any of the submodules in the repository have submodules themselves.

Reply:
I'm not sure I'm following you @chejianj , can  you explain further what you mean? As I understand it, this is what you do at the moment:
* Git clone (not using the `--recurse-submodule` option
* You populate the submodule in the next step (`git submodule update --init`), **note** you're on master at this point
* Conditionally checkout a different tag of the repo. But since you're not using `--recurse-module` here, your submodules will stay at the version that was checked out while you were on master, and that may be different than the one we should use for the tag being checked out.

Reply:
Hi @chejianj , here is a concrete example of what I mean.
* With the current script (not checking out any particular version):
```
$ git describe HEAD --long && git submodule status
ovmf-acrn-v2.3-2-g1c8bef7363
 c3656cc594daac8167721dde7220f0e59ae146fc CryptoPkg/Library/OpensslLib/openssl (OpenSSL_1_1_1d-2-gc3656cc594)
```
* With the current script (checking out 1.1):
```
$ git describe HEAD --long && git submodule status
ovmf-acrn-v1.1-0-gfa06a760f4
+c3656cc594daac8167721dde7220f0e59ae146fc CryptoPkg/Library/OpensslLib/openssl (OpenSSL_1_1_1d-2-gc3656cc594)
```
* With what I am proposing to use (see diff below [1])
```
$ git describe HEAD --long && git submodule status
ovmf-acrn-v1.1-0-gfa06a760f4
 d4e4bd2a8163f355fa8a3884077eaec7adc75ff7 CryptoPkg/Library/OpensslLib/openssl (OpenSSL_1_1_0h)
```

The difference is that in the current version of the script, the version of the submodule stays at the version used in master, although we are checking out an older tag. Does that help understand what I mean?

[1] Proposed changes to the script:
```
diff --git a/doc/static/downloads/build_acrn_ovmf.sh b/doc/static/downloads/build_acrn_ovmf.sh
index e4a5b27d..1a56d504 100644
--- a/doc/static/downloads/build_acrn_ovmf.sh
+++ b/doc/static/downloads/build_acrn_ovmf.sh
@@ -118,23 +118,17 @@ create_acrn_edk2_workspace()
 
     [ -d acrn-edk2 ] && sudo rm -rf acrn-edk2
 
-    git clone https://github.com/projectacrn/acrn-edk2.git
+    git clone --recurse-submodules=CryptoPkg/Library/OpensslLib/openssl https://github.com/projectacrn/acrn-edk2.git
     if [ $? -ne 0 ]; then
         echo "git clone acrn-edk2 failed"
         return 1
     fi
 
     cd acrn-edk2
-    git submodule update --init CryptoPkg/Library/OpensslLib/openssl
-    if [ $? -ne 0 ]; then
-        echo "git submodule acrn-edk2 failed"
-        return 1
-    fi
-
     if [ "${acrn_ver}" != "latest" ]; then
-        git checkout -b "v${acrn_ver}" "ovmf-acrn-v${acrn_ver}"
+        git checkout --recurse-submodules "ovmf-acrn-v${acrn_ver}"
         if [ $? -ne 0 ]; then
-            echo "git checkout -b v${acrn_ver} ovmf-acrn-v${acrn_ver} failed"
+            echo "git checkout --recurse-submodules ovmf-acrn-v${acrn_ver} failed"
             return 1
         fi
     fi
```